blockchain/chaingen/generator.go

```diff
@@ -1716,6 +1716,59 @@ func ReplaceVotes(voteBits uint16, newVersion uint32) func(*wire.MsgBlock) {
 	}
 }
 
+// ReplaceRevocationVersions returns a function that itself takes a block and
+// modifies it by replacing the transaction version of all revocations contained
+// in the block.
+func ReplaceRevocationVersions(newVersion uint16) func(*wire.MsgBlock) {
+	return func(b *wire.MsgBlock) {
+		for _, stx := range b.STransactions {
+			if isRevocationTx(stx) {
+				stx.Version = newVersion
+			}
+		}
+	}
+}
+
+// CreateRevocationsForMissedTickets returns a function that itself takes a
+// block and modifies it by creating revocations for all tickets that would
+// otherwise become missed as of that block.
+func (g *Generator) CreateRevocationsForMissedTickets() func(*wire.MsgBlock) {
+	return func(b *wire.MsgBlock) {
+		// Create a map of the ticket hashes of votes in the block for faster
+		// lookup.
+		const ticketInIdx = 1
+		ticketsPerBlock := g.params.TicketsPerBlock
+		voteTicketHashes := make(map[chainhash.Hash]struct{}, ticketsPerBlock)
+		for _, stx := range b.STransactions {
+			if isVoteTx(stx) {
+				ticketHash := stx.TxIn[ticketInIdx].PreviousOutPoint.Hash
+				voteTicketHashes[ticketHash] = struct{}{}
+			}
+		}
+
+		// Get the winning tickets for the block.
+		winners, _, err := winningTickets(g.tip, g.liveTickets, ticketsPerBlock)
```

Comment:
I believe it's more correct (and generic) to use `g.blocks[b.PrevBlock]` instead of `g.tip` here.

Reply:
Good call.  I agree, though it should always be the tip here since this also relies on the global live tickets state.  I updated this but also added a check to ensure the provided block connects to the current tip.